### ISSUE TITLE
False positive for `void_checks`

### ISSUE THREAD

**Issue by lrhn** [dart-lang/linter#4019]
**Describe the issue**

An assignment of type `Future<dynamic>?` to a parameter of type `FutureOr<void>?` gives 
a "Don't assign to void" warning.

**To Reproduce**
Example:
```dart
  void foo(FutureOr<void>? arg) {}
  Future<dynamic>? future;

  foo(future); // <- lint warning on `future`.
  // Reports:
  // Don't assign to void. dart(void_checks)
```

**Expected behavior**

Don't warn.
If I remove the `Or` from the `FutureOr` of `arg`, I get no warning. 
If I change `Future` to `FutureOr` in the declaration of `future`, I get no warning.
Getting a warning here seems inconsistent.

(I don't know if also disallowing downcast from `dynamic` makes a difference.)

**Additional context**
Occurred while editing `package:async`, so the full set of enabled lints are from:
https://github.com/dart-lang/async/blob/master/analysis_options.yaml
which also includes `strict_casts`.
